### PR TITLE
Change URL Scheme

### LINKED ISSUE #9: Make API endpoints compatible with Prometheus client_go
Using a better URL path would help to integrate with https://github.com/prometheus/client_golang. 
To be more concrete use `/metrics/api/v1/*` scheme instead of `/api/v1/metrics/*`.

Endpoints are defined as constants,
https://github.com/prometheus/client_golang/blob/803ef2a759d7caaaa0de58e3815f1be4c8b5a42a/api/prometheus/v1/api.go#L119-L139

### PR DESCRIPTION
**Pull request by kakkoyun**
This PR changes API URL scheme from `/api/v1/metrics/` to `/api/metrics/v1`, `{interface}/{domain}/{version}`

There are a couple of reasons that justify these changes:
* To be compatible with existing tools such as Prometheus client libraries, e,g `client_go` and Grafana
(These tools expect a base URL and they add paths such as `api/v1/query`)
* To change versions of domains independently, e.g `/api/metrics/v1` and `/api/logs/v3` 

This is a breaking change,  however, Observatorium API GW hasn't been used in production and released a stable version, so this change should be ok.

Resulting APIs that could be use from outside:
```
/api/metrics/v1/api/v1/query
/api/metrics/v1/api/v1/query_range
/api/metrics/v1/write
```

Fixes #9 , #15.

cc @brancz @metalmatze @krasi-georgiev 

# Changes

* Adds a new flags
  -  `--metrics.read.endpoint` fallback base URL for metrics read requests.
    - (Fallback read URL now used to proxy requests that have unrecognized paths directly to backing service.
* Removes `--metrics.query-range.endpoint`, a flag which was a proxy URL to forward query_range requests. It's not needed anymore.

### CODE REVIEW

**Comment by nmagnezi** [comment]
I cherry-picked this PR and build a [test image](https://quay.io/repository/nmagnezi/observatorium?tab=tags) to test thing locally.

I currently get:
```
$ kubectl -n observatorium logs observatorium-xyz-observatorium-api-gateway-78db66cd7c-qckfp
level=info msg="--metrics.read.endpoint is not specified"
msg="parse flag" err="--metrics.query.endpoint and --metrics.query-range.endpoint have to be specified"
```

Probably because this PR dictates a change to the args we should provide to observatorium-api-gateway.

Current params:
```
    spec:
      containers:
      - args:
        - --web.listen=0.0.0.0:8080
        - --metrics.ui.endpoint=http://observatorium-xyz-observatorium-api-gateway-thanos-query.observatorium.svc.cluster.local:9090
        - --metrics.query.endpoint=http://observatorium-xyz-cortex-query-frontend.observatorium.svc.cluster.local:9090/api/v1/query
        - --metrics.write.endpoint=http://observatorium-xyz-thanos-receive.observatorium.svc.cluster.local:19291/api/v1/receive
        - --log.level=warn
        image: quay.io/nmagnezi/observatorium:pr-17-2020-03-15-7458993
```

**Comment by kakkoyun** [comment]
@nmagnezi You have to specify `query` and `query-range` separately. It's already pointed out in the error message.

Could you try it as follows? Thanks in advance, it's great that you're testing it :)

```diff
  observatorium \
    --web.listen=0.0.0.0:8080 \
    --metrics.ui.endpoint=http://127.0.0.1:9091/ \
- # This is not used/needed when metrics.query.endpoint and metrics.query-range.endpoint specified.
- # But better to specify it for now, to support other endpoints that have exposed by Thanos Querier.
+   --metrics.read.endpoint=http://127.0.0.1:9091/api/v1 \
    --metrics.query.endpoint=http://127.0.0.1:9091/api/v1/query \
+   --metrics.query-range.endpoint=http://127.0.0.1:9091/api/v1/query_range \
    --metrics.write.endpoint=http://127.0.0.1:19291/api/v1/receive \
```

**Comment by metalmatze** [comment]
If you say so, I trust you on this rename :)

One thing I want to point out though is that Grafana expecting `api/v1/` is a different API like the one we're versioning. We're talking about Observatorium and then about Prometheus' API and they could theoretically be versioned independently.

**Comment by bwplotka** [comment]
Agree with @metalmatze here.

To me, it might either mean `/v1/metrics/api/v1` or `/metrics/v1/api/v1` (:

**Comment by kakkoyun** [comment]
@metalmatze @bwplotka Good point. I'm going to do it. Thanks.

File: internal/server/server.go [hunk 1]
```diff
@@ -48,36 +48,39 @@ func New(logger log.Logger, reg *prometheus.Registry, opts ...Option) Server {
...
+				ins.newHandler("query", proxy.New(logger, namespace+queryPath, options.metricsQueryEndpoint, options.proxyOptions...)))
+		}
+
+		if options.metricsQueryRangeEndpoint != nil {
+			queryRangePath := "/api/v1/query_range"
+			r.Mount(queryRangePath,
+				ins.newHandler("query_range", proxy.New(logger, namespace+queryRangePath, options.metricsQueryRangeEndpoint, options.proxyOptions...)))
+		}
+
+		if options.metricsReadEndpoint != nil {
```
Comment (brancz) [inline internal/server/server.go]: We don't need to revise this here, but I don't feel like we should differentiate here, just always send this against the same endpoint. The proxy cache will proxy through any requests it doesn't handle, so we can minimize the configuration necessity here.

**Comment by brancz** [comment]
Even if it's slightly odd, it's future proof as it is now, and that's mostly what we were aiming for here, so I'm happy with it.

**Comment by squat** [comment]
Ack I'm happy with it too despite the verbosity. Only one of the versions is actually ours so it's less real complexity than it appears

**Comment by krasi-georgiev** [comment]
agreed, can't think of an alternative either. This suffix is really hard coded in the Prom client and in Grafana.

File: internal/server/server.go (hunk 1 above)
Comment (kakkoyun) [reply internal/server/server.go]: 👍 I'll do the changes.

**Pull request merged**